### PR TITLE
Rescheduling

### PR DESCRIPTION
**Pull request by Firtzberg**
Adds rescheduling decorator.
https://github.com/neighborhoods/KojoWorkerDecoratorComponent/issues/7
[Jira ticket](https://55places.atlassian.net/browse/SERCOMP-2)
Already added a Worker in the [firtness project](https://github.com/neighborhoods/KojoWorkerDecoratorComponentFitness/commit/a8cd82aecd3fcfdc0bb801d99f09c6c0b4ede53d), which is using it and works as intended.

### CODE REVIEW

**Comment by Firtzberg** [comment]
Not sure if this should also be handled as a breaking change.
Earlier none of the predefined decorators had a dependency on throwable-diagnostic-component, but the new rescheduling decorator does. Due to it this additional line is needed when building the container.
``` php
$containerBuilder->addSourcePath(
    '../vendor/neighborhoods/throwable-diagnostic-component/src'
);
```

File: README.md
Comment (Firtzberg) [inline README.md]: also call it the innermost decorator.

**Review by arielallon — APPROVED** [review]
this should be considered a breaking change and be a major version bump.
we'll want a review from a kojo owner before merging.

File: README.md
Comment (mucha55) [inline README.md]: It may be worth linking to the `Fitness` usage of this for clarity: https://github.com/neighborhoods/KojoWorkerDecoratorComponentFitness/blob/main/src/ReschedulingDecorator/Worker/Builder.service.yml#L9

File: src/Worker/ReschedulingDecorator.php
Comment (mucha55) [inline src/Worker/ReschedulingDecorator.php]: Is there any technical reason for this being a strict maximum? I definitely understand making sure an application doesn't use this incorrectly, but kojo doesn't care what this value is, so this seems a little too restrictive. I don't feel strongly about it though, so it's not a blocker

File: src/Worker/ReschedulingDecorator.php
Comment (mucha55) [inline src/Worker/ReschedulingDecorator.php]: Non-blocker: FWIW this can also be done in the constructor like so:

```
new \DateTime(
    sprintf(
        '+%d seconds',
        $this->getRescheduleDelaySeconds()
    )
)
```

File: src/Worker/ReschedulingDecorator/Builder.php
Comment (mucha55) [inline src/Worker/ReschedulingDecorator/Builder.php:30]: **blocker**

This isn't defined by the interface, nor is it used by the decorator

File: src/Worker/ReschedulingDecorator.php
Comment (mucha55) [inline src/Worker/ReschedulingDecorator.php]: **blocker**

Based on the other lines of code I saw (passing in the kojo `RDBMS\Connection\Service`), it looks like you intended to call `setPDO()` somewhere in this decorator, but that doesn't happen. As the code stands now, we don't know if kojo is operating within this transaction (userspace might be setting it but it's not guaranteed)

Comment (mucha55) [inline src/Worker/ReschedulingDecorator.php]: **blocker**

Using `\Doctrine\DBAL\Connection::beginTransaction()` here instead of `\PDO::beginTransaction()` means that there is internal state in userspace (transaction nesting, savepoints, etc.) that kojo does not have access to (if you're interested, you can check out the doctrine code [here](https://github.com/doctrine/dbal/blob/4.0.x/src/Driver/Connection.php) and [here](https://github.com/doctrine/dbal/blob/4.0.x/src/Connection.php) to get a sense of what I'm talking about). It's definitely a pit of failure (though honestly there's not much we can do about it from kojospace), but the simplest way to go about it is to manage transactions at the `\PDO` level (since kojo has a reference to the same `\PDO`, all information is shared across the userspace/kojospace boundary). In LS we go about this by extracting the `\PDO` via `\Doctrine\DBAL\Connection::getWrappedConnection()`, and performing the same `beginTransaction()`, `commit()`, and `rollBack()` operations there

This is a weird and subtle problem, if you're not understanding what I'm talking about, I can put together some examples to demonstrate it (they would be a useful reference to point to in the future anyway)

File: src/Worker/ReschedulingDecorator.php
Comment (mucha55) [inline src/Worker/ReschedulingDecorator.php]: **blocker**

As it stands, a `\Throwable` will mean that no job state change will be applied by this worker, userspace will exit, and kojo will complain, not knowing whether to `complete_success`, `complete_failed`, `panic`, etc. this job

**Review by mucha55 — CHANGES_REQUESTED** [review]
Overall I don't love the fact that an application using this decorator can't do anything to react to a failed re-scheduling, but as long as that fact is made explicit, I don't have a problem with it existing (it's just not what I would use if I did need to be able to react to a failed re-scheduling)

File: src/Worker/ReschedulingDecorator.php
Comment (Firtzberg) [reply src/Worker/ReschedulingDecorator.php]: The `diagnose` method is expected to throw an exception.
This is only one of the decorators wrapping the worker. It's expected that the [ExceptionHandling](https://github.com/neighborhoods/KojoWorkerDecoratorComponent/blob/master/src/Worker/ExceptionHandlingDecorator.php) decorator wraps the others.
We don't want to mix exception handling into this, but let exceptions bubble up. What ReschedulingDecorator does is investigate if the exception is transient or not. When doing so only postgres related diagnostics is applied as indicated [here](https://github.com/neighborhoods/KojoWorkerDecoratorComponent/blob/rescheduling/src/Worker/ReschedulingDecorator/ThrowableDiagnostic/Builder.service.yml#L9).

File: src/Worker/ReschedulingDecorator/Builder.php
Comment (Firtzberg) [reply src/Worker/ReschedulingDecorator/Builder.php:30]: [ReschedulingDecoratorInterface](https://github.com/neighborhoods/KojoWorkerDecoratorComponent/blob/rescheduling/src/Worker/ReschedulingDecoratorInterface.php) extends [DecoratorInterface](https://github.com/neighborhoods/KojoWorkerDecoratorComponent/blob/rescheduling/src/Worker/DecoratorInterface.php), which in turn extends [WorkerInterface](https://github.com/neighborhoods/KojoWorkerDecoratorComponent/blob/rescheduling/src/WorkerInterface.php), which has the `setApiV1RDBMSConnectionService` method.

But it's a good point. It's because the two kojo services get provided by kojo when it runs the worker. We don't have them before and cannot inject them during container building. We also cannot assume if the worker/decorator needs any of them, so we inject them into every one to make them accessible.

I feel like this is the same story as https://github.com/neighborhoods/KojoWorkerDecoratorComponent/issues/4 and could be solved using the proposed approach. The difference is that the the logger isn't kojo specific and therefore doesn't belong in this component, but these two services are so it makes sense to define their placeholders here.
I can make a separate branch to demonstrate how I imagine it if that's not clear.

File: src/Worker/ReschedulingDecorator.php
Comment (Firtzberg) [reply src/Worker/ReschedulingDecorator.php]: The main reason is that we would expect a job to use a cron expression if it runs daily or less frequently.

I could add this into the readme.

File: src/Worker/ReschedulingDecorator.php
Comment (Firtzberg) [reply src/Worker/ReschedulingDecorator.php]: Thank you for the feedback.
The specific kojo connection is configured in the [UserlandPdoDecorator](https://github.com/neighborhoods/KojoWorkerDecoratorComponent#userland-pdo-decorator).
The assumption is that the ReschedulingDecorator is used with the UserlandPdoDecorator and the the UserlandPdoDecorator is the outer. Would stating it explicitly in the readme be enough?

Comment (Firtzberg) [reply src/Worker/ReschedulingDecorator.php]: I see what you mean.
I'll make this `PDO` aware instead of `Doctrine\DBAL\Connection` aware.

Comment (Firtzberg) [reply src/Worker/ReschedulingDecorator.php]: Updates have been made.

File: src/Worker/ReschedulingDecorator.php
Comment (jpmarcotte) [reply src/Worker/ReschedulingDecorator.php]: But if a user wants to and has a valid reason to, is there a technical reason why a user should be prevented from doing so? Shouldn't this set of components be providing flexible and customizable tools for a variety of use-cases and not creating restrictions on how a user should be doing things? From a cron-expression perspective, there's not really any difference between something being weekly, daily, hourly, or every minute, so deciding that clients should exclusively use cron-expressions above one of those thresholds feels arbitrarily restrictive from a user perspective.

File: src/Worker/ReschedulingDecorator.php
Comment (Firtzberg) [reply src/Worker/ReschedulingDecorator.php]: Replaced.

File: src/Worker/ReschedulingDecorator.php
Comment (Firtzberg) [reply src/Worker/ReschedulingDecorator.php]: Limit has been completely removed.

Comment (jpmarcotte) [reply src/Worker/ReschedulingDecorator.php]: Though I did just think of something where someone may want to specify a maximum limit (might be a new decorator) is retrying with an exponential backoff. We do something like that in Listing Service in a few places.

Comment (Firtzberg) [reply src/Worker/ReschedulingDecorator.php]: I think that would go into the Exception Handling Decorator. Currently the exception handling decorator just delays by a constant time in case of a transient exception. I guess you could have different delay strategies and specify which one you want using DI.
That's out of scope for this ticket. Please use the SERCOMP project if you want to add it to the backlog

File: README.md
Comment (Firtzberg) [reply README.md]: Done.

File: src/Worker/ReschedulingDecorator.php
Comment (Firtzberg) [reply src/Worker/ReschedulingDecorator.php]: I have made this more clear in the README.

File: README.md
Comment (mucha55) [inline README.md:114]: I'm curious: do you foresee uses of this decorator that would reschedule a job of a different type?

I know you mentioned earlier that the decorator has no way of determining what the underlying job type is, which is why we need to explicitly specify it, but this is the first I'm hearing of explicitly specifying a different job type

**Review by mucha55 — APPROVED** [review]
Thanks for the clarifications, this looks good to me now

File: README.md
Comment (mucha55) [reply README.md:114]: Relatedly: would it be useful for the kojo worker service API to expose the job type? It would come in useful for a bunch of these decorators, but I don't know if you want to go through the trouble of supporting kojo versions that expose it and those that don't, and having branching logic to get that value

Comment (jpmarcotte) [reply README.md:114]: It could also just have the KWDC 1.1 require Kojo 5.1. I don't think I'd consider that a breaking change.

Comment (Firtzberg) [reply README.md:114]: > I'm curious: do you foresee uses of this decorator that would reschedule a job of a different type?

The intention is rescheduling. But since the user specifies the job type code manually, they could really put anything in there. I cannot think of a concrete use case where the job type code would differ from the current job, mainly because I cannot imagine a job using this decorator if it isn't an observer.
If there were use cases where the job type code can differ, this decorator would not _reschedule_, but simply _schedule_. So it's name is inappropriate.

Would you like me to made a change to the readme before merging?

**Pull request merged**
